README.md
# AWS IoT SDK for JavaScript
The aws-iot-device-sdk.js package allows developers to write JavaScript 
applications which access the AWS IoT Platform via [MQTT or MQTT over the Secure WebSocket Protocol](http://docs.aws.amazon.com/iot/latest/developerguide/protocols.html).  It can be used in Node.js environments as well as in browser applications.

* [Overview](#overview)
* [Installation](#install)
* [Examples](#examples)
* [API Documentation](#api)
* [Connection Types](#connections)
* [Example Programs](#programs)
* [Browser Applications](#browser)
* [Troubleshooting](#troubleshooting)
* [Unit Tests](#unittests)
* [License](#license)
* [Support](#support)

<a name="overview"></a>
## Overview
This document provides instructions on how to install and configure the AWS 
IoT device SDK for JavaScript, and includes examples demonstrating use of the
SDK APIs.

### MQTT Connection
This package is built on top of [mqtt.js](https://github.com/mqttjs/MQTT.js/blob/master/README.md) and provides three classes: 'device', 'thingShadow' and 'jobs'.  The 'device' class wraps [mqtt.js](https://github.com/mqttjs/MQTT.js/blob/master/README.md) to provide a
secure connection to the AWS IoT platform and expose the [mqtt.js](https://github.com/mqttjs/MQTT.js/blob/master/README.md) interfaces upward.  It provides features to simplify handling of intermittent connections, including progressive backoff retries, automatic re-subscription upon connection, and queued offline publishing with configurable drain rate.

### Collection of Metrics
Beginning with Release v2.2.0 of the SDK, AWS collects usage metrics indicating which language and version of the SDK is being used. This allows us to prioritize our resources towards addressing issues faster in SDKs that see the most and is an important data point. However, we do understand that not all customers would want to report this data by default. In that case, the sending of usage metrics can be easily disabled by set options.enableMetrics to false.

### Thing Shadows
The 'thingShadow' class implements additional functionality for accessing Thing Shadows via the AWS IoT
API; the thingShadow class allows devices to update, be notified of changes to,
get the current state of, or delete Thing Shadows from AWS IoT.  Thing
Shadows allow applications and devices to synchronize their state on the AWS IoT platform.
For example, a remote device can update its Thing Shadow in AWS IoT, allowing
a user to view the device's last reported state via a mobile app.  The user
can also update the device's Thing Shadow in AWS IoT and the remote device 
will synchronize with the new state.  The 'thingShadow' class supports multiple 
Thing Shadows per mqtt connection and allows pass-through of non-Thing-Shadow
topics and mqtt events.

### Jobs
The 'jobs' class implements functionality to interact with the AWS IoT Jobs service. The
IoT Job service manages deployment of IoT fleet wide tasks such as device software/firmware 
deployments and updates, rotation of security certificates, device reboots, and custom device
specific management tasks. 

Included in this package is an example 'agent'. The agent can be used either as a stand-alone
program to manage installation and maintenance of files and other running processes or it can 
be incorporated into a customized agent to meet specific application needs.

<a name="install"></a>
## Installation

**NOTE:** AWS IoT Node.js SDK will only support Node version 4 or above. 

You can check your node version by 

```sh
node -v
```

Installing with npm:

```sh
npm install aws-iot-device-sdk
```

Installing from github:

```sh
git clone https://github.com/aws/aws-iot-device-sdk-js.git
cd aws-iot-device-sdk-js
npm install
```

<a name="examples"></a>
## Examples

### Device Class
```js
var awsIot = require('aws-iot-device-sdk');

//
// Replace the values of '<YourUniqueClientIdentifier>' and '<YourCustomEndpoint>'
// with a unique client identifier and custom host endpoint provided in AWS IoT.
// NOTE: client identifiers must be unique within your AWS account; if a client attempts 
// to connect with a client identifier which is already in use, the existing 
// connection will be terminated.
//
var device = awsIot.device({
   keyPath: <YourPrivateKeyPath>,
  certPath: <YourCertificatePath>,
    caPath: <YourRootCACertificatePath>,
  clientId: <YourUniqueClientIdentifier>,
      host: <YourCustomEndpoint>
});

//
// Device is an instance returned by mqtt.Client(), see mqtt.js for full
// documentation.
//
device
  .on('connect', function() {
    console.log('connect');
    device.subscribe('topic_1');
    device.publish('topic_2', JSON.stringify({ test_data: 1}));
  });

device
  .on('message', function(topic, payload) {
    console.log('message', topic, payload.toString());
  });
```
### Thing Shadow Class
```js
var awsIot = require('aws-iot-device-sdk');

//
// Replace the values of '<YourUniqueClientIdentifier>' and '<YourCustomEndpoint>'
// with a unique client identifier and custom host endpoint provided in AWS IoT cloud
// NOTE: client identifiers must be unique within your AWS account; if a client attempts 
// to connect with a client identifier which is already in use, the existing 
// connection will be terminated.
//
var thingShadows = awsIot.thingShadow({
   keyPath: <YourPrivateKeyPath>,
  certPath: <YourCertificatePath>,
    caPath: <YourRootCACertificatePath>,
  clientId: <YourUniqueClientIdentifier>,
      host: <YourCustomEndpoint>
});

//
// Client token value returned from thingShadows.update() operation
//
var clientTokenUpdate;

//
// Simulated device values
//
var rval = 187;
var gval = 114;
var bval = 222;

thingShadows.on('connect', function() {
//
// After connecting to the AWS IoT platform, register interest in the
// Thing Shadow named 'RGBLedLamp'.
//
    thingShadows.register( 'RGBLedLamp', {}, function() {

// Once registration is complete, update the Thing Shadow named
// 'RGBLedLamp' with the latest device state and save the clientToken
// so that we can correlate it with status or timeout events.
//
// Thing shadow state
//
       var rgbLedLampState = {"state":{"desired":{"red":rval,"green":gval,"blue":bval}}};

       clientTokenUpdate = thingShadows.update('RGBLedLamp', rgbLedLampState  );
//
// The update method returns a clientToken; if non-null, this value will
// be sent in a 'status' event when the operation completes, allowing you
// to know whether or not the update was successful.  If the update method
// returns null, it's because another operation is currently in progress and
// you'll need to wait until it completes (or times out) before updating the 
// shadow.
//
       if (clientTokenUpdate === null)
       {
          console.log('update shadow failed, operation still in progress');
       }
    });
});
thingShadows.on('status', 
    function(thingName, stat, clientToken, stateObject) {
       console.log('received '+stat+' on '+thingName+': '+
                   JSON.stringify(stateObject));
//
// These events report the status of update(), get(), and delete() 
// calls.  The clientToken value associated with the event will have
// the same value which was returned in an earlier call to get(),
// update(), or delete().  Use status events to keep track of the
// status of shadow operations.
//
    });

thingShadows.on('delta', 
    function(thingName, stateObject) {
       console.log('received delta on '+thingName+': '+
                   JSON.stringify(stateObject));
    });

thingShadows.on('timeout',
    function(thingName, clientToken) {
       console.log('received timeout on '+thingName+
                   ' with token: '+ clientToken);
//
// In the event that a shadow operation times out, you'll receive
// one of these events.  The clientToken value associated with the
// event will have the same value which was returned in an earlier
// call to get(), update(), or delete().
//
    });

```
### Jobs Class
```js
var awsIot = require('aws-iot-device-sdk');

//
// Replace the values of '<YourUniqueClientIdentifier>' and '<YourCustomEndpoint>'
// with a unique client identifier and custom host endpoint provided in AWS IoT cloud
// NOTE: client identifiers must be unique within your AWS account; if a client attempts 
// to connect with a client identifier which is already in use, the existing 
// connection will be terminated.
//
var jobs = awsIot.jobs({
   keyPath: <YourPrivateKeyPath>,
  certPath: <YourCertificatePath>,
    caPath: <YourRootCACertificatePath>,
  clientId: <YourUniqueClientIdentifier>,
      host: <YourCustomEndpoint>
});

//
// Jobs is built on top of awsIot.device and inherits all of the same functionality.
//
jobs
  .on('connect', function() {
    console.log('connect');
    device.subscribe('topic_1');
    device.publish('topic_2', JSON.stringify({ test_data: 1}));
    });

jobs
  .on('message', function(topic, payload) {
    console.log('message', topic, payload.toString());
  });

//
// To subscribe to job execution events call the subscribeToJobs method which takes
// a callback that will be invoked when a job execution is available or an error 
// occurs. The job object passed to the callback contains information about the job
// execution and methods for updating the job execution status. Details covered 
// in the API documentation below.
//
jobs.subscribeToJobs(thingName, function(err, job) { 
   if (isUndefined(err)) {
      console.log('default job handler invoked, jobId: ' + job.id.toString()); 
      console.log('job document: ' + job.document);
   }
   else {
      console.error(err);
   }
});

jobs.subscribeToJobs(thingName, 'customJob', function(err, job) { 
   if (isUndefined(err)) {
      console.log('customJob operation handler invoked, jobId: ' + job.id.toString()); 
      console.log('job document: ' + job.document);
   }
   else {
      console.error(err);
   }
});

//
// After calling subscribeToJobs for each operation on a particular thing call
// startJobNotifications to cause any existing queued job executions for the given
// thing to be published to the appropriate subscribeToJobs handler. Only needs
// to be called once per thing.
//
jobs.startJobNotifications(thingName, function(err) {
   if (isUndefined(err)) {
      console.log('job notifications initiated for thing: ' + thingName);
   }
   else {
      console.error(err);
   }
});

```

<a name="api"></a>
## API Documentation

  * <a href="#device"><code>awsIot.<b>device()</b></code></a>
  * <a href="#thingShadow"><code>awsIot.<b>thingShadow()</b></code></a>
  * <a href="#jobs"><code>awsIot.<b>jobs()</b></code></a>
  * <a href="#register"><code>awsIot.thingShadow#<b>register()</b></code></a>
  * <a href="#unregister"><code>awsIot.thingShadow#<b>unregister()</b></code></a>
  * <a href="#update"><code>awsIot.thingShadow#<b>update()</b></code></a>
  * <a href="#get"><code>awsIot.thingShadow#<b>get()</b></code></a>
  * <a href="#delete"><code>awsIot.thingShadow#<b>delete()</b></code></a>
  * <a href="#publish"><code>awsIot.thingShadow#<b>publish()</b></code></a>
  * <a href="#subscribe"><code>awsIot.thingShadow#<b>subscribe()</b></code></a>
  * <a href="#unsubscribe"><code>awsIot.thingShadow#<b>unsubscribe()</b></code></a>
  * <a href="#end"><code>awsIot.thingShadow#<b>end()</b></code></a>
  * <a href="#subscribeToJobs"><code>awsIot.jobs#<b>subscribeToJobs()</b></code></a>
  * <a href="#unsubscribeFromJobs"><code>awsIot.jobs#<b>unsubscribeFromJobs()</b></code></a>
  * <a href="#startJobNotifications"><code>awsIot.jobs#<b>startJobNotifications()</b></code></a>
  * <a href="#job"><code><b>job</b></code></a>
  * <a href="#document"><code>job#<b>document</b></code></a>
  * <a href="#id"><code>job#<b>id</b></code></a>
  * <a href="#operation"><code>job#<b>operation</b></code></a>
  * <a href="#status"><code>job#<b>status</b></code></a>
  * <a href="#inProgress"><code>job#<b>inProgress()</b></code></a>
  * <a href="#failed"><code>job#<b>failed()</b></code></a>
  * <a href="#succeeded"><code>job#<b>succeeded()</b></code></a>

-------------------------------------------------------
<a name="device"></a>
### awsIot.device(options)

Returns a wrapper for the [mqtt.Client()](https://github.com/mqttjs/MQTT.js/blob/master/README.md#client) 
class, configured for a TLS connection with the AWS IoT platform and with 
arguments as specified in `options`.  The AWSIoT-specific arguments are as 
follows:

  * `host`: the AWS IoT endpoint you will use to connect
  * `clientId`: the client ID you will use to connect to AWS IoT
  * `certPath`: path of the client certificate file
  * `keyPath`: path of the private key file associated with the client certificate
  * `caPath`: path of your CA certificate file 
  * `clientCert`: same as `certPath`, but can also accept a buffer containing client certificate data
  * `privateKey`: same as `keyPath`, but can also accept a buffer containing private key data
  * `caCert`: same as `caPath`, but can also accept a buffer containing CA certificate data
  * `autoResubscribe`: set to 'true' to automatically re-subscribe to topics after reconnection (default 'true')
  * `offlineQueueing`: set to 'true' to automatically queue published messages while offline (default 'true')
  * `offlineQueueMaxSize`: enforce a maximum size for the offline message queue (default 0, e.g. no maximum)
  * `offlineQueueDropBehavior`: set to 'oldest' or 'newest' to define drop behavior on a full queue when offlineQueueMaxSize > 0
  * `drainTimeMs`: the minimum time in milliseconds between publishes when draining after reconnection (default 250)
  * `baseReconnectTimeMs`: the base reconnection time in milliseconds (default 1000)
  * `maximumReconnectTimeMs`: the maximum reconnection time in milliseconds (default 128000)
  * `minimumConnectionTimeMs`: the minimum time in milliseconds that a connection must be maintained in order to be considered stable (default 20000)
  * `protocol`: the connection type, either 'mqtts' (default), 'wss' (WebSocket/TLS), or 'wss-custom-auth' (WebSocket/TLS with custom authentication).  Note that when set to 'wss', values must be provided for the Access Key ID and Secret Key in either the following options or in environment variables as specified in [WebSocket Configuration](#websockets). When set to 'wss-custom-auth', valid headers must be provided as specified in [Custom Auth](#custom-auth)
  * `websocketOptions`: if `protocol` is set to 'wss', you can use this parameter to pass additional options to the underlying WebSocket object; these options are documented [here](https://github.com/websockets/ws/blob/master/doc/ws.md#class-wswebsocket).
  * `filename`: used to load credentials from the file different than the default localtion when `protocol` is set to 'wss'. Default value is '~/.aws/credentials'
  * `profile`: used to specify which credential profile to be used when `protocol` is set to 'wss'.  Default value is 'default'
  * `accessKeyId`: used to specify the Access Key ID when `protocol` is set to 'wss'.  Overrides the environment variable `AWS_ACCESS_KEY_ID` and `AWS_ACCESS_KEY_ID` from `filename` if set.
  * `secretKey`: used to specify the Secret Key when `protocol` is set to 'wss'.  Overrides the environment variable `AWS_SECRET_ACCESS_KEY`and `AWS_SECRET_ACCESS_KEY`  from `filename` if set.
  * `sessionToken`: (required when authenticating via Cognito, optional otherwise) used to specify the Session Token when `protocol` is set to 'wss'.  Overrides the environment variable `AWS_SESSION_TOKEN` if set.
  * `customAuthHeaders`: used to specify your custom authorization headers when `protocol` is set to 'wss-custom-auth'. The fields 'X-Amz-CustomAuthorizer-Name', 'X-Amz-CustomAuthorizer-Signature', and the field for your token name are required.
  * `keepalive`: used to specify the time interval for each ping request. Default is set to 300 seconds to connect to AWS IoT.
  * `enableMetrics`: used to report SDK version usage metrics. It is set to true by default. To disable metrics collection, set value to false.


CHANGELOG.md
## [2.2.2](https://github.com/aws/aws-iot-device-sdk-js/releases/tag/v2.2.2) (Sep 24, 2019)

Bugfixes/Imporovements
  - Reorder timeout state cleanup in order to support invoking update from a timeout callback

## [2.2.1](https://github.com/aws/aws-iot-device-sdk-js/releases/tag/v2.2.1) (Jan 24, 2018)

Bugfixes/Imporovements
  - Upgrade MQTT.js to 2.15.1 to address security issue

## [2.2.0](https://github.com/aws/aws-iot-device-sdk-js/releases/tag/v2.2.0) (Nov 29, 2017)

Features
  - Added AWS IoT Job API
  - Added options to enable AWS IoT Custom Auth
  - Added options to enalbe/disable AWS IoT metrics collection
  - Added new API to support packetsend and packetreceive events

Bugfixes/Imporovements
  - Modify Keepalive defaults to 300 secs to maintain consistency across SDKs
  - Expose shadow version from raw json object
  - Added samples to demonstrate AWS IoT Job API
  - Disabled MQTT.js default resubscribe. 

## [2.1.0](https://github.com/aws/aws-iot-device-sdk-js/releases/tag/v2.0.1) (Sep 28, 2017)

Features
  - Update MQTT.js to 2.13.0. [MQTT.js](https://github.com/mqttjs/MQTT.js/releases/tag/v2.13.0)

Bugfixes/Imporovements
  - Propagated 'error' from 'close' event. [#131](https://github.com/aws/aws-iot-device-sdk-js/pull/131)
  - Fixed method of handleMessage to be overridden rather than pass-through. [#129](https://github.com/aws/aws-iot-device-sdk-js/pull/129)
  - Pass 'connack' parameter in 'connect' event  [#99](https://github.com/aws/aws-iot-device-sdk-js/pull/99)
  - Update iot service name to 'iotdevicegateway'

## [2.0.1](https://github.com/aws/aws-iot-device-sdk-js/releases/tag/v2.0.1) (Jul 2, 2017)

Bugfixes/Imporovements
  - Removed validation against .com in websocket connection.

## [2.0.0](https://github.com/aws/aws-iot-device-sdk-js/releases/tag/v2.0.0) (Mar 21, 2017)

API Changes
  - Deprecated region option(-g) in device configuration.
  - Added host endpoint option(-H) to connect to custom host endpoint 

Features
  - Added support for browserify on Windows CMD. [#74](https://github.com/aws/aws-iot-device-sdk-js/issues/74)
  - Added support for loading IAM credentials from aws credential files.
  - Added sample for using Node.js SDK with webpack.

Bugfixes/Imporovements
  - Fixed README.md typo [#101](https://github.com/aws/aws-iot-device-sdk-js/issues/101)
  - Fixed thing.register() API to have independent optional parameters.[#106](https://github.com/aws/aws-iot-device-sdk-js/issues/106) 
  - Upgrade MQTT.js to v2.2.1 and gulp dependencies. 
  - Fixed npm test failure in node version above 4. 

## [1.0.14](https://github.com/aws/aws-iot-device-sdk-js/releases/tag/v1.0.14) (Dec 7, 2016)

Bugfixes/Improvements
  - Fixes for GitHub issues [#67]( https://github.com/aws/aws-iot-device-sdk-js/issues/67), [#95](https://github.com/aws/aws-iot-device-sdk-js/issues/95), [#96](https://github.com/aws/aws-iot-device-sdk-js/issues/96).

## [1.0.13](https://github.com/aws/aws-iot-device-sdk-js/releases/tag/v1.0.13) (July 11, 2016)

Bugfixes/Improvements
  - Addressed pull request[#83](https://github.com/aws/aws-iot-device-sdk-js/pull/83) Credits given to [Torsph](https://github.com/Torsph)
  - Updated lifecycle events browser demo to read from DynamoDB table of connected clients if available
  - Addresses pull request [#60](https://github.com/aws/aws-iot-device-sdk-js/pull/60)
  - Fixes for GitHub issues [#66](https://github.com/aws/aws-iot-device-sdk-js/issues/66), [#61](https://github.com/aws/aws-iot-device-sdk-js/issues/61), [#53](https://github.com/aws/aws-iot-device-sdk-js/issues/53), [#48](https://github.com/aws/aws-iot-device-sdk-js/issues/48), and [#44](https://github.com/aws/aws-iot-device-sdk-js/issues/44).

## [1.0.12](https://github.com/aws/aws-iot-device-sdk-js/releases/tag/v1.0.12) (April 19, 2016)

Features
  - Added support for use in browser applications

Bugfixes/Improvements
  - Incorporated GitHub pull request [#49](https://github.com/aws/aws-iot-device-sdk-js/pull/49)
  - Fixes for GitHub issues [#41](https://github.com/aws/aws-iot-device-sdk-js/issues/41), [#47](https://github.com/aws/aws-iot-device-sdk-js/issues/47), and [#50](https://github.com/aws/aws-iot-device-sdk-js/issues/50).

## [1.0.11](https://github.com/aws/aws-iot-device-sdk-js/releases/tag/v1.0.11) (March 4, 2016)

Features:
  - Configurable exponential backoff retries after connection loss
  - Configurable offline publish message queueing
  - Added option for automatic re-subscription after reconnect
  - Added shadow option for versioning disable
  - Added session token support

Bugfixes/Improvements
  - Incorporated github pull requests [#33](https://github.com/aws/aws-iot-device-sdk-js/pull/33), [#34](https://github.com/aws/aws-iot-device-sdk-js/pull/34), and [#39](https://github.com/aws/aws-iot-device-sdk-js/pull/39)
  - Fixes for github issue [#36](https://github.com/aws/aws-iot-device-sdk-js/issues/36)
  - Updated unit tests
  - Updated documentation

## [1.0.10](https://github.com/aws/aws-iot-device-sdk-js/releases/tag/v1.0.10) (January 28, 2016)

Features:
  - Added support for WebSocket connections to AWS IoT

Bugfixes/Improvements
  - Incorporated github pull requests [#28](https://github.com/aws/aws-iot-device-sdk-js/pull/28) and [#29](https://github.com/aws/aws-iot-device-sdk-js/pull/29)
  - Fixes for github issues [#30](https://github.com/aws/aws-iot-device-sdk-js/issues/30)
  - Added unit tests to release
  - Updated documentation

## [1.0.7](https://github.com/aws/aws-iot-device-sdk-js/releases/tag/v1.0.7) (October 30, 2015)

Bugfixes/Improvements:
  - Incorporated github pull requests [#7](https://github.com/aws/aws-iot-device-sdk-js/pull/7), [#9](https://github.com/aws/aws-iot-device-sdk-js/pull/9), and [#14.](https://github.com/aws/aws-iot-device-sdk-js/pull/14)
  - Fixes for github issues [#8](https://github.com/aws/aws-iot-device-sdk-js/issues/8) and [#16.](https://github.com/aws/aws-iot-device-sdk-js/issues/16)
  - Updated documentation
  - JSHint cleanup


examples/temperature-control.js
/*
 * Copyright 2010-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

//node.js deps

//npm deps

//app deps
const blessed = require('blessed');
const contrib = require('blessed-contrib');
const thingShadow = require('../thing');
const isUndefined = require('../common/lib/is-undefined');
const cmdLineProcess   = require('./lib/cmdline');


function processTest( args, argsRemaining ) {

//
// Construct user interface
//
var screen = blessed.screen( { enableInput: true });

screen.title = 'AWS IoT Temperature Control Simulation';

var grid = new contrib.grid( { rows: 8, cols: 12, screen: screen } );

//
// The temperature control device's control state contains:
//
//   setPoint - If enabled, heat will be transferred between the interior
//              and exterior until the interior temperature equals this
//              value
//
//              Type: Integer
//              Units: Degrees Fahrenheit
//              Read/Write
//
//    enabled - Operational status of heat transfer device
//
//              Type: Boolean
//              Units: N/A
//              Read/Write
var deviceControlState={ setPoint: 72, enabled: true };

//
// The temperature control device's monitor state contains:
//
//
//    intTemp - Interior temperature
//
//              Type: Integer
//              Units: Degrees Fahrenheit
//              Read Only
//
//    extTemp - Exterior temperature
//
//              Type: Integer
//              Units: Degrees Fahrenheit
//              Read Only
//
//   curState - Current state of heat transfer device
//
//              Type: Enum (heating|cooling|stopped)
//              Units: N/A
//              Read Only
//
var deviceMonitorState={ intTemp: 72, extTemp: 45, curState: 'stopped' };

var networkEnabled=true;

var lcd1 = grid.set( 1, 1, 2, 3, contrib.lcd, {
label: "Setpoint",
    segmentWidth: 0.06,
    segmentInterval: 0.11,
    strokeWidth: 0.1,
    elements: 3,
    elementSpacing: 4,
    elementPadding: 2, color: 'green'
 
} );

var lcd2 = grid.set( 3, 1, 2, 3, contrib.lcd, { 
label: "Interior",
    segmentWidth: 0.06,
    segmentInterval: 0.11,
    strokeWidth: 0.1,
    elements: 3,
    elementSpacing: 4,
    elementPadding: 2, color: 'white'
} );

var lcd3 = grid.set( 5, 1, 2, 3, contrib.lcd, { 
label: "Exterior",
    segmentWidth: 0.06,
    segmentInterval: 0.11,
    strokeWidth: 0.1,
    elements: 3,
    elementSpacing: 4,
    elementPadding: 2, color: 'white'
} );

var lcd4 = grid.set( 2, 4, 2, 3, contrib.lcd, { 
label: "Status",
    segmentWidth: 0.06,
    segmentInterval: 0.11,
    strokeWidth: 0.1,
    elements: 3,
    elementSpacing: 4,
    elementPadding: 2, color: 'white'
} );

var lcd5 = grid.set( 4, 4, 2, 3, contrib.lcd, { 
label: "Network",
    segmentWidth: 0.06,
    segmentInterval: 0.11,
    strokeWidth: 0.1,
    elements: 3,
    elementSpacing: 4,
    elementPadding: 2, color: 'white'
} );

var log = grid.set( 1, 7, 6, 5, contrib.log, {
label: "Log",
} );



screen.key(['escape', 'q', 'C-c'], function(ch, key) {
  return process.exit(0);
});


lcd1.setDisplay(deviceControlState.setPoint+'F');
lcd2.setDisplay(deviceMonitorState.intTemp+'F');
lcd3.setDisplay(deviceMonitorState.extTemp+'F');
lcd4.setDisplay(deviceControlState.enabled===true?' ON':'OFF');
lcd5.setDisplay(networkEnabled===true?' ON':'OFF');

screen.render()

//
// The thing module exports the thing class through which we
// can register and unregister interest in thing shadows, perform
// update/get/delete operations on them, and receive delta updates
// when the cloud state differs from the device state.
//
const thingShadows = thingShadow({
  keyPath: args.privateKey,
  certPath: args.clientCert,
  caPath: args.caCert,
  clientId: args.clientId,
  region: args.region,
  reconnectPeriod: args.reconnectPeriod,
});

var opClientToken;

var title;
if (args.testMode === 2)
{
title = blessed.text( {
   top: 0,
   left: 'center',
   align: 'center',
   content: 'DEVICE SIMULATOR'
} );
}
else
{
title = blessed.text( {
   top: 0,
   left: 'center',
   align: 'center',
   content: 'MOBILE APPLICATION SIMULATOR'
} );
}
var help = blessed.text( {
   top: 1,
   left: 'center',
   align: 'center',
   content: '(use arrow keys to change temperature)'
} );

var bar = blessed.listbar({
  //parent: screen,
  bottom: 0,
  left: 'center',
//  right: 3,
  height: 'true' ? 'shrink' : 3,
//  mouse: true,
  keys: true,
  autoCommandKeys: false,
  border: 'line',
  vi: true,
  style: {
    bg: 'black',
    item: {
      bg: 'black',
      hover: {
        bg: 'blue'
      },
      //focus: {
      //  bg: 'blue'
      //}
    },
    selected: {
      bg: 'blue'
    }
  },
  commands: {
    'mode': {
      callback: function() {
        var enabledStatus=( deviceControlState.enabled===true?'OFF':' ON');
        deviceControlState.enabled=(deviceControlState.enabled===true?false:true);
        log.log('temperature control '+(deviceControlState.enabled?'enabled':'disabled'));

        if (networkEnabled===true)
        {
           opClientToken = thingShadows.update('TemperatureControl', { state: { desired: deviceControlState } });
        }
        lcd4.setDisplay( enabledStatus );
        screen.render();
      }
    },
    'network': {
      callback: function() {
        var networkStatus=( networkEnabled===true ? 'OFF' : ' ON');
        networkEnabled=( networkEnabled===true ? false : true);
        lcd5.setDisplay( networkStatus );
        log.log( 'network '+(networkEnabled?'connected':'disconnected'));
//
// Simulate a network connection/disconnection 
//
        if (networkEnabled)
        {
           thingShadows.setConnectionStatus(true);

//
// After re-connecting, try to update with our current state; this will
// get a 'rejected' status if another entity has updated this thing shadow
// in the meantime and we will have to re-sync.
//
           thingShadows.update( 'TemperatureControl', { state: { desired: deviceControlState } } );
        }
        else
        {
           thingShadows.setConnectionStatus(false);
        }
        screen.render();
      }
    },
    'exit': {
      callback: function() {
        process.exit(0);
      }
    }
  }
});

screen.append(bar);
screen.append(help);
screen.append(title);

bar.focus();


screen.key('up', function( ch, key ) {
if (deviceControlState.setPoint < 90)
{
   deviceControlState.setPoint++;
   lcd1.setDisplay(deviceControlState.setPoint+'F');
   if (networkEnabled===true)
   {
      opClientToken = thingShadows.update('TemperatureControl', { state: { desired: deviceControlState } });
   }
   screen.render()
}
});
screen.key('down', function( ch, key ) {
if (deviceControlState.setPoint > 50)
{
   deviceControlState.setPoint--;
   lcd1.setDisplay(deviceControlState.setPoint+'F');
   if (networkEnabled===true)
   {
      opClientToken = thingShadows.update('TemperatureControl', { state: { desired: deviceControlState } });
   }
   screen.render()
}
});

//
// Simulate the interaction of a mobile application and a remote device via the
// AWS IoT service.  The simulated remote device is a temperature controller.
// Two thing shadows are used: one for control properties (setpoint and on/off) 
// and the other for monitoring properties (interior/exterior temperature, current
// state).
//
thingShadows
  .on('connect', function() {
    log.log('connected to AWS IoT...');

    thingShadows.register( 'TemperatureControl', { 
                                              persistentSubscribe: true } );
    thingShadows.register( 'TemperatureStatus', { 
                                              persistentSubscribe: true } );
    log.log('registered thing shadows...');
//
// After registering, wait for a few seconds and then ask for the
// current state of the thing shadow.
//
    setTimeout( function() {
       opClientToken = thingShadows.get('TemperatureControl');
    }, 2000 );
    });

thingShadows 
  .on('close', function() {
    thingShadows.unregister( 'TemperatureControl' );
    log.log('close');
  });

thingShadows 
  .on('reconnect', function() {
    log.log('reconnect/re-register');
    thingShadows.register( 'TemperatureControl', { 
                                              persistentSubscribe: true } );
//
// After re-registering, wait for a few seconds and then try to update
// with our current state
//
    setTimeout( function() {
       opClientToken = thingShadows.update('TemperatureControl', { state: { desired: deviceControlState } });
    }, 2000 );
  });

thingShadows 
  .on('offline', function() {
    log.log('offline');
  });

thingShadows
  .on('error', function(error) {
    log.log('error', error);
  });

thingShadows
  .on('message', function(topic, payload) {
    log.log('message', topic, payload.toString());
  });

thingShadows
  .on('status', function(thingName, statusType, clientToken, stateObject) {
      if ((networkEnabled===true) && (statusType === 'rejected'))
      {
//
// If an operation is rejected it is likely due to a version conflict;
// request the latest version so that we synchronize with the thing
// shadow.  The most notable exception to this is if the thing shadow
// has not yet been created or has been deleted.
//
         if (stateObject.code !== 404)
         {
            log.log('resync with thing shadow');
            opClientToken = thingShadows.get(thingName);
         }
      }
      if (statusType === 'accepted')
      {
         if (thingName==='TemperatureControl')
         {
            deviceControlState = stateObject.state.desired;
            lcd1.setDisplay(deviceControlState.setPoint+'F');
            lcd4.setDisplay(deviceControlState.enabled===true?' ON':'OFF');
            screen.render()
         }
      }
  });

thingShadows
  .on('delta', function(thingName, stateObject) {
      if (networkEnabled===true)
      {
         if (thingName==='TemperatureControl')
         {
            if (!isUndefined(stateObject.state.enabled) && 
                (stateObject.state.enabled !== deviceControlState.enabled))
            {
               log.log('temperature control '+(stateObject.state.enabled?'enabled':'disabled'));
            }
            deviceControlState=stateObject.state;
            lcd1.setDisplay(deviceControlState.setPoint+'F');
            lcd4.setDisplay(deviceControlState.enabled===true?' ON':'OFF');
         }
         else if (thingName==='TemperatureStatus')
         {
            if (!isUndefined(stateObject.state.intTemp))
            {
               deviceMonitorState.intTemp=stateObject.state.intTemp;
            }
            if (!isUndefined(stateObject.state.extTemp))
            {
               deviceMonitorState.extTemp=stateObject.state.extTemp;
            }
            if (!isUndefined(stateObject.state.curState))
            {
               deviceMonitorState.curState=stateObject.state.curState;
            }
         }
         screen.render()
      }
  });

//
// If we are acting in the device simulation role, we simulate the
// action of the heat transfer device here.
//
   if (args.testMode === 2)
   {
      var transferRate = 0.1;  // degrees Fahrenheit/second
      var leakRate = 0.005;     // degrees Fahrenheit/second
      var interiorTemp = deviceMonitorState.intTemp;

      setInterval( function() {
         var difference;
         var logMsg = '';
//
// If the device is enabled, the internal temperature will move towards
// the setpoint; otherwise, it will move towards the external temperature.
//
         if (deviceControlState.enabled)
         {
            difference=deviceMonitorState.intTemp-deviceControlState.setPoint;
         }
         else
         {
            difference=deviceMonitorState.intTemp-deviceMonitorState.extTemp;
         }
         interiorTemp -= (difference * transferRate);
         interiorTemp -= leakRate;
         deviceMonitorState.intTemp = Math.floor(interiorTemp);
         if (deviceControlState.enabled)
         {
            if (difference > transferRate)
            {
               deviceMonitorState.curState = 'cooling';
            }
            else if (difference < -transferRate)
            {
               deviceMonitorState.curState = 'heating';
            }
            else
            {
               deviceMonitorState.curState = 'stopped';
            }
         }
         else
         {
            deviceMonitorState.curState = 'stopped';
         }
         if (networkEnabled===true)
         {
            opClientToken = thingShadows.update('TemperatureStatus', 
                                { state: { desired: deviceMonitorState } });
         }
      }, 1000 );
   }
//
// The interior temperature LCD also displays the status of the heat
// transfer device.
//
   setInterval( function() {
      var displayOptions = { color: 'white' };
      if (deviceMonitorState.curState === 'cooling')
      {
         displayOptions = { color: 'cyan' };
      }
      else if (deviceMonitorState.curState === 'heating')
      {
         displayOptions = { color: 'red' };
      }
      lcd2.setDisplay(deviceMonitorState.intTemp+'F');
      lcd3.setDisplay(deviceMonitorState.extTemp+'F');
      lcd2.setOptions( displayOptions );
      screen.render();
   }, 250 );
}

module.exports = cmdLineProcess;

if (require.main === module) {
  cmdLineProcess('connect to the AWS IoT service and demonstrate thing shadow APIs, test modes 1-2',
                 process.argv.slice(2), processTest )
}
